### PR TITLE
SONAR-7986 Remove property set values when removing property set

### PR DESCRIPTION
**Pull request by julienlancelot**
SONAR-7986 Remove property set values when removing property set

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsUpdater.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+      return;
+    }
+    Set<String> propertySetKeys = extractPropertySetKeys(propertyDto.get(), definition);
+    for (String key : propertySetKeys) {
+      deleteSetting(dbSession, key, componentDto);
+    }
+    deleteSetting(dbSession, propertyKey, componentDto);
+  }
+
+  private Optional<PropertyDto> selectPropertyDto(DbSession dbSession, String propertyKey, Optional<ComponentDto> componentDto) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsUpdater.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsUpdater.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  public void deleteGlobalSetting(DbSession dbSession, String propertyKey) {
+    delete(dbSession, propertyKey, Optional.empty());
+  }
+
+  public void deleteComponentSetting(DbSession dbSession, String propertyKey, ComponentDto componentDto) {
+    delete(dbSession, propertyKey, Optional.of(componentDto));
+  }
+
+  private void delete(DbSession dbSession, String propertyKey, Optional<ComponentDto> componentDto) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsUpdater.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsUpdater.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+  private void deleteSetting(DbSession dbSession, String propertyKey, Optional<ComponentDto> componentDto) {
+    if (componentDto.isPresent()) {
+      dbClient.propertiesDao().deleteProjectProperty(propertyKey, componentDto.get().getId(), dbSession);
+    } else {
+      dbClient.propertiesDao().deleteGlobalProperty(propertyKey, dbSession);
+    }
+  }
+
+  private void deletePropertySet(DbSession dbSession, String propertyKey, PropertyDefinition definition, Optional<ComponentDto> componentDto) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsUpdater.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsUpdater.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  private void delete(DbSession dbSession, String propertyKey, Optional<ComponentDto> componentDto) {
+    PropertyDefinition definition = definitions.get(propertyKey);
+    if (definition == null || !definition.type().equals(PropertyType.PROPERTY_SET)) {
+      deleteSetting(dbSession, propertyKey, componentDto);
+    } else {
+      deletePropertySet(dbSession, propertyKey, definition, componentDto);
+    }
+  }
+
+  private void deleteSetting(DbSession dbSession, String propertyKey, Optional<ComponentDto> componentDto) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsUpdater.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* MINOR 4 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/PropertySetExtractor.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+  private static final Splitter COMMA_SPLITTER = Splitter.on(",");
+
+  private PropertySetExtractor() {
+    // Only static stuff
+  }
+
+  public static Set<String> extractPropertySetKeys(PropertyDto propertyDto, PropertyDefinition definition) {
+    Set<String> propertySetKeys = new HashSet<>();
+    definition.fields()
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/PropertySetExtractor.java]: suggestion: sometimes, 2 `for` loop are clearer than 2 `forEach`. In your case, it's not complex, but I had a use case where it was clearer with for loops

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/PropertySetExtractor.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      .forEach(field -> COMMA_SPLITTER.splitToList(propertyDto.getValue())
+        .forEach(setId -> propertySetKeys.add(generatePropertySetKey(propertyDto.getKey(), setId, field.key()))));
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/PropertySetExtractor.java]: that's when I realize I had the property set format wrong : )

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsFinder.java [hunk 7]
```diff
@@ -87,16 +87,10 @@ public SettingsFinder(DbClient dbClient, PropertyDefinitions definitions) {
     properties.stream()
       .filter(propertyDto -> definitions.get(propertyDto.getKey()) != null)
       .filter(propertyDto -> definitions.get(propertyDto.getKey()).type().equals(PROPERTY_SET))
-      .forEach(propertyDto -> definitions.get(propertyDto.getKey()).fields()
-        .forEach(field -> COMMA_SPLITTER.splitToList(propertyDto.getValue())
-          .forEach(setId -> propertySetKeys.add(generatePropertySetKey(propertyDto.getKey(), setId, field.key())))));
+      .forEach(propertyDto -> propertySetKeys.addAll(extractPropertySetKeys(propertyDto, definitions.get(propertyDto.getKey()))));
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsFinder.java]: what about a `flatMap(property -> extractPropertySetKey(property, definitions.get(property.getKey())))` and `collect(Collectors.toSet())` in the end?

File: server/sonar-server/src/test/java/org/sonar/server/settings/ws/ResetActionTest.java [hunk 8]
```diff
@@ -93,7 +97,7 @@ public void remove_global_setting() throws Exception {
 
   @Test
   public void remove_component_setting() throws Exception {
-    setUserAsSystemAdmin();
+    setUserAsProjectAdmin();
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/settings/ws/ResetActionTest.java]: a unit test is missing on property set, no?

File: it/it-tests/src/test/java/it/settings/PropertySetsTest.java [hunk 9]
```diff
@@ -22,76 +22,125 @@
...
-  static String getProperty(String key) {
-    return orchestrator.getServer().getAdminWsClient().find(new PropertyQuery().setKey(key)).getValue();
+  static void resetSetting(String... keys){
+    for (String key : keys){
+      settingsService().reset(ResetRequest.builder().setKey(key).build());
+    }
+  }
+
+  private static SettingsService settingsService() {
+    return newAdminWsClient(orchestrator).settingsService();
```
Comment (teryk) [inline it/it-tests/src/test/java/it/settings/PropertySetsTest.java]: each time you call `settingsService` you create a new admin ws client. I suggest to have the admin ws client as a parameter

File: it/it-tests/src/test/java/it/settings/PropertySetsTest.java [hunk 10]
```diff
@@ -22,76 +22,125 @@
...
+   */
+  @Deprecated
   static void setProperty(String key, String value) {
     orchestrator.getServer().getAdminWsClient().update(new PropertyUpdateQuery(key, value));
   }
 
-  static String getProperty(String key) {
-    return orchestrator.getServer().getAdminWsClient().find(new PropertyQuery().setKey(key)).getValue();
+  static void resetSetting(String... keys){
+    for (String key : keys){
```
Comment (teryk) [inline it/it-tests/src/test/java/it/settings/PropertySetsTest.java]: suggestion: here you could use `keys.forEach(…)`

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsFinder.java (hunk 7 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsFinder.java]: Indeed, it should work, thanks !

**Comment by teryk** [comment]
check suggestions and…
lgtm-jesus

File: server/sonar-server/src/test/java/org/sonar/server/settings/ws/ResetActionTest.java (hunk 8 above)
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/settings/ws/ResetActionTest.java]: I've added them in the SettingsUpdater

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/PropertySetExtractor.java (hunk 6 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/PropertySetExtractor.java]: oups

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsFinder.java (hunk 7 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsFinder.java]: For the record, the right way is : 
```
      .flatMap(propertyDto -> extractPropertySetKeys(propertyDto, definitions.get(propertyDto.getKey())).stream())
      .collect(Collectors.toSet());
```

**Pull request merged**